kedro/io/core.py

```diff
@@ -148,6 +159,23 @@ class AbstractDataset(abc.ABC, Generic[_DI, _DO]):
     need to change the `_EPHEMERAL` attribute to 'True'.
     """
     _EPHEMERAL = False
+    # Declares a class-level attribute that will store the initialization
+    # arguments of an instance. Initially, it is set to None, but it will
+    # hold a dictionary of arguments after initialization.
+    _init_args: dict[str, Any] | None = None
```

Comment:
Isn't this shared across all instances? How to make sure we're not overwriting everything with the latest instance args?

Reply:
Since we override it later on in post init:

```
self._init_args = call_args
```

it becomes an instance attribute when we access it through an instance. So this can only be a problem if someone accesses it through class `ClassName._init_args`.

Technically, we don't need it there. But since `AbstractDataset` is an abstract class and doesn't have a constructor, I declared it as a class attribute.

Reply:
Test example

```python
from inspect import getcallargs
from typing import Callable, Any


class Parent:
    _init_args: dict[str, Any] | None = None

    def __init_subclass__(cls, **kwargs):
        init_func: Callable = cls.__init__

        def init_decorator(previous_init: Callable) -> Callable:
            def new_init(self, *args, **kwargs) -> None:
                previous_init(self, *args, **kwargs)
                if type(self) is cls:
                    call_args = getcallargs(init_func, self, *args, **kwargs)
                    self.__post_init__(call_args)

            return new_init

        cls.__init__ = init_decorator(cls.__init__)

    def __post_init__(self, call_args: dict[str, Any]) -> None:
        self._init_args = call_args
        self._init_args.pop("self", None)


class Child(Parent):
    def __init__(self, z, a=2, b=3):
        super().__init__()
        print('Child __init__')


class ChildAnother(Parent):
    def __init__(self, z, a=2, b=3):
        super().__init__()
        print('ChildAnother __init__')


class GrandChild(Child):
    def __init__(self, z, c=5, d=5):
        super().__init__(z)
        print('GrandChild __init__')


def main():
    child = Child(z=1, a=2, b=3)
    child_another = ChildAnother(z=10, a=20, b=30)
    grandchild = GrandChild(z=4, c=5, d=6)

    print(child._init_args)
    print(child_another._init_args)
    print(grandchild._init_args)


if __name__ == "__main__":
    main()
```

Output

```console
Child __init__
ChildAnother __init__
Child __init__
GrandChild __init__
{'z': 1, 'a': 2, 'b': 3}
{'z': 10, 'a': 20, 'b': 30}
{'z': 4, 'c': 5, 'd': 6}

```

Reply:
I added a comment about overriding

Reply:
I don't think we need it at all and it might introduce a confusion, having an  class property and instance property with the same name. The wrapper set in `__init_subclass__` will be the one adding the instance property, so it should all be fine.